src/monitor/node_active_protocol.c

```diff
@@ -1873,38 +1924,43 @@ set_node_candidate_priority(PG_FUNCTION_ARGS)
 			GetPrimaryNodeInGroup(currentNode->formationId,
 								  currentNode->groupId);
 
-		if (primaryNode == NULL)
+		/*
+		 * If we allow setting changes during APPLY_SETTINGS we open the door
+		 * for race conditions where we can't be sure that the latest changes
+		 * have been applied.
+		 *
+		 * If we don't currently have a primary node anyway, we can just
+		 * proceed with the change.
+		 */
+		if (primaryNode &&
+			!IsCurrentState(primaryNode, REPLICATION_STATE_APPLY_SETTINGS))
```

Comment:
What happens now if we change the settings in WAIT_PRIMARY? Doesn't this have the same issue as doing it during APPLY_SETTINGS?

Reply:
Not really, not anymore. Nowadays the APPLY_SETTINGS to either WAIT_PRIMARY or PRIMARY is computed in the `ProceedGroupStateForPrimaryNode` function in the Group FSM, in such a way that the state that was observed before entering APPLY_SETTINGS is not important anymore. Things might change either during the operation, or because of the new settings (e.g. number sync standbys).

The only case we don't know how to handle properly is concurrent calls to change the replication settings, and that mostly is a client side problem, where we want to return from the `pg_autoctl` to the shell (or caller) only when the setting is known to have been applied.